### ISSUE TITLE
Restrict state transitions according to graph

### ISSUE THREAD

**Issue by ericelliott** [ericelliott/redux-dsm#5]
Should switch reducers based on the current state in order to ensure that the program can't enter bad states.
- [x] For each state, build a reducer that only responds to transitions listed for that state in the state graph
- [ ] Create a parent/meta reducer that switches the invoked child reducer based on the current state

Sample state graph:

``` js
const fetchingStates = [
  ['initialize', 'idle',
    ['fetch', 'fetching',
      ['cancel', 'idle'],
      ['report error', 'error',
        ['handle error', 'idle']
      ],
      ['report success', 'success',
        ['handle success', 'idle']
      ]
    ]
  ]
];
```

**Comment by zouhir** [comment]
I'll look at this tonight, first wanna make sure I 100% understand what I have to do

**Comment by Rewieer** [comment]
I'm not quite sure how is this supposed to be implemented.

**Comment by dash-** [comment]
I've been looking at this for an hour or so, and I think I get it, but I'd like to bounce my understanding off of @ericelliott and make sure this is right.

In the example:
1. We start off with state = "idle".  I can see that in the code as being the default state.
2. From "idle", we can only process "initialize" and "fetch" actions.
3. From "fetching", we can only process "cancel", "report error", and "report success" actions.
4. From "error", we can only process "handle error" actions.
5. From "success", we can only process "handle success" actions.

Said more generally, the reducer exposed by dsm() should immediately just return the current redux application state (and not invoke "sub-reducers") when the action given is not an immediate child of the current DSM state (in the DSM state graph).

So, if the current DSM state is "initialized", the DSM reducer shouldn't process "cancel" actions.  After all, there's nothing to cancel at the moment.

All of that being said, I'd actually recommend a slightly different example graph for the real-world:

```javascript
const fetchingStates = [
  ['initialize', 'initialized',
    ['fetch', 'fetching',
      ['cancel', 'initialized'],
      ['report error', 'error',
        ['handle error', 'initialized']
      ],
      ['report success', 'success',
        ['handle success', 'initialized']
      ]
    ]
  ]
];
```

Do I understand this correctly?  If so, I can definitely make this happen.

**Comment by dash-** [comment]
Also, assuming that my understanding is correct, why build a bunch of reducers w/ a parent reducer for switching between them?  Wouldn't it be more straightforward to just modify the function returned by createReducer to bail early if it detects that the current DSM state does not include the action type (within an allowed transitions map)?

For example:

```javascript
const createReducer = (defaultState, reducerMap) => {
  return (state = defaultState, action = {}) => {
    const { type } = action;

    if (allowedTransitions[state].includes(type)) {
      if (typeof reducer === 'function') {
        return reducer(state, action);
      }
    } else {
      // Handle error condition somehow
    }

    return state;
  };
```

Note that I am assuming that detecting an illegal state transition should throw an error or somehow inform the developer/testers that something undesirable has occurred.  It is, after all, probably an indication that they have an issue somewhere in their logic.  Otherwise, if no indications of this issue are given, I can see some very difficult to debug situations arising.  Although I'm not completely decided on how this should be handled.

Thoughts?

**Comment by ericelliott** [comment]
If you do the if statement version, it will have to search the entire allowedTransitions array every time there's an action. In other words, that's a less efficient implementation than the one I had in mind. Basically, I want something like this to happen:

```js
const initReducer = (state = {data: {}, status: 'IDLE'}, action) => {
  switch (action.type) {
    case 'FETCH':
      return Object.assign({}, state, {
        // transition to a new state
        status: 'FETCHING'
      });
    default:
      return state;
  }
};

const fetchingReducer = (state = {data: {}, status: 'FETCHING'}, action) => {
  switch (action.type) {
  case 'CANCEL':
    return Object.assign({}, state, {
      // transition to a new state
      status: 'IDLE'
    });
  case 'REPORT_ERROR':
    return Object.assign({}, state, {
      error: action.payload.error,
      status: 'ERROR'
    });
  case 'REPORT_SUCCESS': 
    return Object.assign({}, state, {
      data: action.payload,
      status: 'SUCCESS'
    });
  default:
    return state;
  }
};


const fetchReducer = (state, action) => {
    switch (state.status) {
    case 'INIT':
        return initReducer(state, action);
    case 'FETCHING':
        return fetchingReducer(state, action);
    default:
        return initReducer(state, action)
    }
};
```

Ideally, I want to generate these reducers automatically -- this is a lot of boilerplate just to handle some async behaviors. The idea of redux-dsm is to eliminate this boilerplate and present a more declarative way to create these reducers.

**Comment by dash-** [comment]
Okay, I think I get it completely.  Let me think on it, and maybe I'll attack it tomorrow.  Thanks for your time and help understanding this!

**Comment by ericelliott** [comment]
Published as v1.0.3. Thanks @dash- !
